Fix evented pleg mirror pod & use IsEventedPLEGInUse instead of FG status check

pacoxu opened the pull request:
This PR is based on #122763, and we use `IsEventedPLEGInUse()` instead of `utilfeature.DefaultFeatureGate.Enabled(features.EventedPLEG)`.

- This PR also includes #127954 which can fix init container problem.

Fixes #123087

Context can be found in #122763.

```release-note
kubelet: fix a static pod startup bug with Evented PLEG
```

pacoxu committed:
for CREATED state, skip if EventedPLEG is enabled

pacoxu committed:
fix eventedPLEG flake: init container may waiting with empty reason before seeting it to PodInitializing

pacoxu committed:
add log for whether trigger start/restart for Created state

pacoxu commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-kubernetes-e2e-kind-beta-features

pacoxu commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features

pacoxu commented:
/test pull-kubernetes-unit
/test pull-kubernetes-integration

pacoxu commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features

pacoxu commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features

pacoxu commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features

pacoxu commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features

pacoxu commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features

pacoxu commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features

pacoxu commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features

pacoxu commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features

pacoxu committed:
use IsEventedPLEGInUse instead of check FG status

pacoxu commented:
/test pull-kubernetes-e2e-kind-beta-features
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e

pacoxu commented:
In last commit https://github.com/kubernetes/kubernetes/pull/122778/commits/3135122e122562ea02d4abc8bbcc4d676232fdff, I tried to use `IsEventedPLEGInUse()` to avoid using FG status in code base. 

Everything works well util now.

/cc @aojea @harche

harche commented:
/cc @sairameshv

pacoxu commented:
/kind bug

pacoxu commented:
/test pull-kubernetes-e2e-kind-beta-features
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e

p0lyn0mial commented:
/test pull-kubernetes-e2e-gce-cos-alpha-features

harche commented:
I tried to create a reproducer by messing around with Generic PLEG and syncPod to get the container stay in [created state](https://github.com/kubernetes/kubernetes/blob/909faa3a9b33161992af0684c5582353b5be6d9b/pkg/kubelet/container/runtime.go#L266) but ended up blowing up my cluster. I am still optimistic though. We should be able to have a stable reproducer. When you look at the logs from the runtime and kubelet, you can see how quickly the container moves from created to running state. I am trying to invoke syncPod while the container is still at created state.   

I will continue to look having a stable enough reproducer setup.

pacoxu commented:
/milestone v1.30

harche commented on pkg/kubelet/container/cache.go:
```diff
@@ -103,12 +101,11 @@ func (c *cache) Set(id types.UID, status *PodStatus, err error, timestamp time.T
-		if cachedVal, ok := c.pods[id]; ok && cachedVal.modified.After(timestamp) {
-			return false
-		}
+	// Added after Evented PLEG was added
```
@pacoxu should it be gated by pleg.IsEventedPLEGInUse() (since we are removing `utilfeature.DefaultFeatureGate.Enabled(features.EventedPLEG)` and it was added after Evented PLEG?

harche replied:
It will also maintain the existing behavior intact for Generic PLEG.

pacoxu replied:
there will be an import cycle 😓

harche replied:
In that case, IMO it would make sense to keep `utilfeature.DefaultFeatureGate.Enabled(features.EventedPLEG)`  because otherwise it alters the behavior in existing Generic PLEG and we don't want to risk that.

harche commented on pkg/kubelet/container/helpers.go:
```diff
@@ -86,10 +86,17 @@ func ShouldContainerBeRestarted(container *v1.Container, pod *v1.Pod, podStatus
+	if status.State == ContainerStateUnknown {
 		return true
 	}
+	if status.State == ContainerStateCreated {
```
Although it is likely that Generic PLEG might also have the same issue with `ShouldContainerBeRestarted` function, but maybe it is not getting surfaced. But considering it is somehow working with Generic PLEG as is, do you want to surround this change as well with `pleg.IsEventedPLEGInUse()` so it would not change code flow for existing Generic PLEG? 

This way the PR we could be merged as it is not modifying the existing behavior under Generic PLEG and help us  unblock the periodic blocking alpha features job. 

We can leave a comment here that, before Evented PLEG is graduated back to Beta, we should have a concrete answer on why the `ShouldContainerBeRestarted()` is reached when using Evented PLEG but not when using Generic PLEG. 

What do you think?

pacoxu replied:
Agree. 

However to avoid import cycle, I add logic when calling `ShouldContainerBeRestarted`. https://github.com/kubernetes/kubernetes/pull/122778/files#diff-0e1a8d22b0f1dc42d412749c6a916b4c76394ead739d4bfbf26564faedd74a2fR971 if evented PLEG is in use, set restartOnCreatedState to false and if not set it true.

harche commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -980,6 +981,9 @@ func (m *kubeGenericRuntimeManager) computePodActions(ctx context.Context, pod *
 						reason: reasonUnknown,
 					}
 				}
+			} else if containerStatus.State == kubecontainer.ContainerStateCreated {
```
Maybe surround the change with `pleg.IsEventedPLEGInUse()`? https://github.com/kubernetes/kubernetes/pull/122778/files#r1459564818

pacoxu replied:
let me update it

pacoxu replied:
Updated.

aojea commented:
please get ACK from @smarterclayton before merge

pacoxu commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-kubernetes-e2e-kind-beta-features

pacoxu commented on pkg/kubelet/container/cache.go:
```diff
@@ -104,6 +104,7 @@ func (c *cache) Set(id types.UID, status *PodStatus, err error, timestamp time.T
 	defer c.lock.Unlock()
 
 	if utilfeature.DefaultFeatureGate.Enabled(features.EventedPLEG) {
+		// Added after Evented PLEG was added
```
This is the only place except `kubelet.go` to calling the FG status. See https://github.com/kubernetes/kubernetes/pull/122778#discussion_r1459966444 for why we keep it.

- Using `pleg.InUse()` will make a import cycle.

harche replied:
/hold 

In light of this PR, https://github.com/kubernetes/kubernetes/pull/120942/files, I am wondering if we need the code within this if block at all. 

I will run some jobs removing this code and get back here. 

https://github.com/kubernetes/kubernetes/pull/122737/files#diff-b010d1c9ff51c8975875be97aeebd3c8c6206128f520115752f9dc007e8afcd1R104

harche replied:
@pacoxu we can remove this block altogether,

https://github.com/kubernetes/kubernetes/pull/122737/files#diff-b010d1c9ff51c8975875be97aeebd3c8c6206128f520115752f9dc007e8afcd1R104

`pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e` [passes](https://github.com/kubernetes/kubernetes/pull/122737#issuecomment-1904283564) fine without that block.

pacoxu replied:
Removed.

harche replied:
Thanks. 

/hold cancel

pacoxu commented:
/assign @smarterclayton
for review

This is based on https://github.com/kubernetes/kubernetes/pull/122763. The key difference is https://github.com/kubernetes/kubernetes/pull/122778/commits/6d4c1dd2fd299f2b2c86dec9a4f852fdafe88f43 which use `pleg.IsEvenetedPLEGInUse()` instead of `DefaultFeatureGate.Enabled(features.EventedPLEG)`

pacoxu commented:
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-kubernetes-e2e-kind-beta-features
/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e
/test pull-kubernetes-e2e-gce-cos-alpha-features

harche reviewed (commented):
/lgtm

harche commented:
/test pull-kubernetes-e2e-kind-evented-pleg

harche commented:
/test pull-kubernetes-e2e-kind-evented-pleg

harche commented:
/test pull-kubernetes-e2e-kind-evented-pleg

harche commented:
/test pull-kubernetes-e2e-kind

harche commented:
> @pacoxu: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-kubernetes-e2e-kind-evented-pleg	[43ff990](https://github.com/kubernetes/kubernetes/commit/43ff9900187ff544f1b0cdccb90f78f9a15e59fa)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/122778/pull-kubernetes-e2e-kind-evented-pleg/1752814076678901760)	false	`/test pull-kubernetes-e2e-kind-evented-pleg`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=122778). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Apacoxu). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

The job has [timeout of 60m](https://github.com/kubernetes/test-infra/blob/23c1cae7f0005a054131c41b067427edc9b0ca5d/config/jobs/kubernetes/sig-testing/kubernetes-kind.yaml#L441) but from the [logs](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/122778/pull-kubernetes-e2e-kind-evented-pleg/1752814076678901760) it seems it went over it and got killed because of the timeout. Maybe we need to bump up the timeout.

pacoxu commented:
/test pull-kubernetes-e2e-kind-evented-pleg

for timeout

harche commented:
@pacoxu not sure why `pull-kubernetes-e2e-kind-evented-pleg` taking this long to run. Maybe something isn't right in the job configuration. I remember before [disabling](https://github.com/kubernetes/test-infra/pull/31647) the Evented PLEG in `pull-kubernetes-e2e-kind-alpha-features`, that job used to pass for this PR. 

I just raised https://github.com/kubernetes/test-infra/pull/31802, to run `pull-kubernetes-e2e-kind-evented-pleg`  more like `pull-kubernetes-e2e-kind-alpha-features`, hopefully that helps.

pacoxu commented:
@harche  https://github.com/kubernetes/kubernetes/pull/122927#issuecomment-1923490257 the presubmit CI failed. Does this presubmit has some other issues?

harche commented:
> @harche [#122927 (comment)](https://github.com/kubernetes/kubernetes/pull/122927#issuecomment-1923490257) the presubmit CI failed. Does this presubmit has some other issues?

Failure to run the job that targets all api could be due to some misconfiguration of [pull-kubernetes-e2e-kind-evented-pleg](https://github.com/kubernetes/test-infra/blob/8f023a0da6de8b3afe50283b562427b1843a0e9b/config/jobs/kubernetes/sig-testing/kubernetes-kind.yaml#L430) job or some underlying issue in the feature that is getting surfaced. 

However, if we want to take a step by step approach, then in my opinion we should first stabilize this feature at the alpha level. Existing alpha features job such as, `pull-kubernetes-e2e-kind-alpha-features`, target only [{"api/alpha":"true", "api/ga":"true"}](https://github.com/kubernetes/test-infra/blob/8f023a0da6de8b3afe50283b562427b1843a0e9b/config/jobs/kubernetes/sig-testing/kubernetes-kind.yaml#L312C19-L312C56). This is the reason I raised a test-infra [PR](https://github.com/kubernetes/test-infra/pull/31802/files#diff-4726678f319d035781356b12f259f830e89db912def662c393971444dbe5a128R456) to make sure `pull-kubernetes-e2e-kind-evented-pleg` also targets same subset of api. 

When we see Evented PLEG being stable enough at the alpha and would consider promoting it to beta, at that time we can modify `pull-kubernetes-e2e-kind-evented-pleg` to target[ '{"api/beta":"true", "api/ga":"true"}'](https://github.com/kubernetes/test-infra/blob/8f023a0da6de8b3afe50283b562427b1843a0e9b/config/jobs/kubernetes/sig-testing/kubernetes-kind.yaml#L562) just like the beta features job does.

harche commented:
/test pull-kubernetes-e2e-kind-evented-pleg

harche commented:
> /test pull-kubernetes-e2e-kind-evented-pleg

@pacoxu after merging https://github.com/kubernetes/test-infra/pull/31802, that job passes now.

pacoxu commented:
/test pull-kubernetes-e2e-kind-evented-pleg

sairameshv commented:
/test pull-kubernetes-e2e-kind-evented-pleg

MaryamTavakkoli commented:
Hello!
Release Signal shadow here.
The code freeze is starting [02:00 UTC Wednesday 6th March 2024 / 18:00 PDT Tuesday 5th March 2024](https://everytimezone.com/s/5056bace) (about three weeks from now), and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As this PR is tagged for 1.30, is it still planned for this release?

pacoxu commented:
> As this PR is tagged for 1.30, is it still planned for this release?

PTAL @smarterclayton @dchen1107 when you have time.

pacoxu commented:
/priority important-soon

pacoxu commented:
cc @yujuhong 
as Yu-ju may be familiar with this, according to https://github.com/kubernetes/kubernetes/issues/123057#issuecomment-1958404200.

pacoxu commented:
Hello!
on behalf of Release Signal Team.

The code freeze is starting [02:00 UTC Wednesday 6th March 2024 / 18:00 PDT Tuesday 5th March 2024](https://everytimezone.com/s/5056bace) in a week, and we want to ensure that each PR has a chance to be merged on time.

pacoxu commented:
We missed the code freeze of v1.30. 
Ping @smarterclayton if you have time to take a look.
We can later cherry-pick to old releases then.

liggitt commented:
/milestone clear

pacoxu commented:
/remove-lifecycle stale
@smarterclayton @yujuhong Do you have time to take a look in v1.31 release cycle?

aojea commented:
/assign @yujuhong

pacoxu commented:
/test pull-kubernetes-e2e-kind-beta-features
/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e
/test pull-kubernetes-e2e-kind-evented-pleg

wxx213 commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -965,7 +966,9 @@ func (m *kubeGenericRuntimeManager) computePodActions(ctx context.Context, pod *
-			if kubecontainer.ShouldContainerBeRestarted(&container, pod, podStatus) {
+			// when Evented PLEG is in use, don't restart
+			// for Generic PLEG, restart as is
+			if kubecontainer.ShouldContainerBeRestarted(&container, pod, podStatus, !pleg.IsEventedPLEGInUse()) {
```
When the container is in `Created `state, the new container with the same name will not be created here, are you assuming that the runtime will start it later? When kubelet call cri function `StartContainer`, the container will be in `Created` state, this will make the container stuck in `Created` state forever.

pacoxu replied:
Can we write an e2e test for this scenario?

Is there a simple way to reproduce the `startContainer timeout`  in my env?

pacoxu replied:
If this is a problem, how about adding a `timeout` check here to see if the pod is in `CREATED` state for a long time? I will try to update the PR later if I can reproduce it.

wxx213 replied:
> Can we write an e2e test for this scenario?
> 
> Is there a simple way to reproduce the `startContainer timeout` in my env?

I added some delay codes in containerd to make the `StartContainer` rpc timeout, then the container stuck in `Created `state.

The patch code in [containerd](https://github.com/containerd/containerd/blob/release/1.7/pkg/cri/server/container_start.go#L57) is  like:
```go
if meta.Config != nil {
		for k,v := range meta.Config.Labels {
			if strings.Compare(k, "io.kubernetes.pod.name") == 0 {
				if strings.Compare(v, "centos7-pod") == 0 {
					if blockStat == 0 {
						blockStat = 1
						time.Sleep(5 * time.Minute)
						return nil, errors.New("timeout")
					}
				}
			}
		}
	}
```

and the pod yaml:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: centos7-pod
  labels:
    name: centos7-pod
spec:
  containers:
  - name: centos7-pod
    image: centos:7
    env:
    - name: GET_HOST_FROM
      value: env
    ports:
    - containerPort: 80
    command:
    - sleep
    args:
    - infinity
  nodeSelector:
    kubernetes.io/hostname: ${your node}
```

pacoxu replied:
Thanks.

I use a mock cni to just `sleep` and reproduce the issue that you mentioned. 

/hold

I will try to address this issue.

pacoxu replied:
A quick fix can be found with the latest commit https://github.com/kubernetes/kubernetes/pull/122778/commits/f09c70d7aa5244dedda084bfcd67955e29abd97f.

@wxx213 could you help to test with the lastest code? I will try it later.

pacoxu replied:
I tested locally with your change in containerd.

With the new commit, 
```
centos7-pod   0/1     ContainerCreating   0               24s
centos7-pod   0/1     RunContainerError   0               2m3s
centos7-pod   1/1     Running             1               2m3s
```

Without the new commit(enable EventedPLEG)
```
centos7-pod   0/1     ContainerCreating   0               12s
centos7-pod   0/1     RunContainerError   0               2m13s
...
centos7-pod   0/1     RunContainerError   0               10m 
```

This fixed the problem that mentioned here.

wxx213 commented:
@pacoxu Is this fixing issue only for mirror pod?

pacoxu committed:
restart with a time check for evented pleg: default 1m

pacoxu commented:
/test pull-kubernetes-e2e-kind-beta-features
/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e
/test pull-kubernetes-e2e-kind-evented-pleg

wxx213 commented:
> We don't restart just because the CREATE state is noticed by EventedPLEG too quick.

@pacoxu Could you give more detail for the issue that this pr trying fix? how the  quickly EventedPLEG notice caused it?

It seems that you are fixing a container "dual running" issue,  I believe this is a classic challenge for kubelet will always hard to know the real state of the container when the rpc timeout appears.

pacoxu commented:
> > We don't restart just because the CREATE state is noticed by EventedPLEG too quick.
> 
> @pacoxu Could you give more detail for the issue that this pr trying to fix? how the quickly EventedPLEG notice caused it?
> 
> It seems that you are fixing a container "dual running" issue, I believe this is a classic challenge for kubelet will always hard to know the real state of the container when the rpc timeout appears.

The `dual running` issue is not caused by rpc timeout, but for a immediate restart(recreate) for a new container of a static pod currently, this doesn't happen for pod that is not a static pod in the kubernetes manifest. 
This can be reproduced in the e2e CI using kind. apiserver/controller-manager will fail to start due to dup container created for the same pod, and the second will fail to start due to port already binding. More details in https://github.com/kubernetes/kubernetes/issues/123087.

pacoxu commented:
/test pull-kubernetes-integration

pacoxu commented:
> > The `dual running` issue is not caused by rpc timeout, but for a immediate restart(recreate) for a new container of a static pod currently, this doesn't happen for pod that is not a static pod in the kubernetes manifest. This can be reproduced in the e2e CI using kind. apiserver/controller-manager will fail to start due to dup container created for the same pod, and the second will fail to start due to port already binding. More details in #123087.
> 
> @pacoxu I am temporarily unable to access kubernets slack for your discussion context, I'm still confused that how the dup container was created, as we know from the codes, only one `podWorkerLoop` work could be executed per pod where the pod containers are created and started, could you give your assumption for this?

@wxx213 This only happened to static pod. In the test CI failure, we find the apiserver pod cannot be started correctly. This bug doesn't occur to a pod which is not static pod, and this is why some users is using the EventedPLEG in their production without a problem. 

This bug should be triggered when below two things happened at the same time.
- for the change here, a `CREATED` event from container runtime triggered a `restart` here.
 - if it is not EventedPLEG, but Generic PLEG, the `CREATED` state of a static pod is in a quite short period and will be ignored.
 - if it is not a staticpod, the logic may be like your statement: the pod should be handled by a podWorkerLoop. 
- meanwhile, the old `CREATED` container is started(probably, the startContainer is called at the same time the restart happens above.)

BTW, I tried to reproduce the failure in https://github.com/kubernetes/kubernetes/pull/127063 to show the problem.

pacoxu commented:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/127063/pull-kubernetes-e2e-kind-alpha-features/1830544265197916160

@wxx213 You can see the dup container in https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/127063/pull-kubernetes-e2e-kind-alpha-features/1830544265197916160/artifacts/kind-control-plane/pods/kube-system_kube-apiserver-kind-control-plane_6c276a318e8e1532ab2e7dd8dbcfebba/kube-apiserver/ 

The kubelet log can be found in https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/127063/pull-kubernetes-e2e-kind-alpha-features/1830544265197916160/artifacts/kind-control-plane/

wxx213 commented:
> https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/127063/pull-kubernetes-e2e-kind-alpha-features/1830544265197916160
> 
> @wxx213 You can see the dup container in https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/127063/pull-kubernetes-e2e-kind-alpha-features/1830544265197916160/artifacts/kind-control-plane/pods/kube-system_kube-apiserver-kind-control-plane_6c276a318e8e1532ab2e7dd8dbcfebba/kube-apiserver/
> 
> The kubelet log can be found in https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/127063/pull-kubernetes-e2e-kind-alpha-features/1830544265197916160/artifacts/kind-control-plane/

Thanks, I checked the kubelet log for pod kube-system/kube-apiserver-kind-control-plane, and it seems that the start event of the  app container wasn't synced to kubelet by evented pleg timely caused this issue. It's a little same as the rpc timeout situation  where sometimes the pod is trapped in  "dual running" state that what we are suffering.

I summarized the timeline from the log for pod kube-system/kube-apiserver-kind-control-plane:
```
I0902 10:03:22.220403  kubelet.go:1758] "SyncPod enter", the first sync loop for pod kube-system/kube-apiserver-kind-control-plane started
I0902 10:03:22.710694  evented.go:295 "Received Container Started Event", the evented pleg got container start event, 
    container id: 5dd8a71e5d80f85e0a7c672b7c9a865807c595e403aefa8a33d7266cf5d92811 which is a sandbox container

I0902 10:03:22.710935  pod_workers.go:963] "Notifying pod of pending update, then triggered a pod work pendding for pod 
    kube-system/kube-apiserver-kind-control-plane

I0902 10:03:23.634078  evented.go:292 "Received Container Created Event" , the evented pleg got container created event, 
    container id: eacee74973c186af7e16228a0fb22bb624a25ade9c499e5355910c8659a53968 which is a app container

I0902 10:03:23.839023  kubelet.go:1764 "SyncPod exit", the first sync loop for pod kube-system/kube-apiserver-kind-control-plane finished

I0902 10:03:23.839162  kubelet.go:1758] "SyncPod enter", the second sync loop for pod kube-system/kube-apiserver-kind-control-plane started, the work only 
    got the created state from evented pleg of the app container eacee74973c186af7e16228a0fb22bb624a25ade9c499e5355910c8659a53968
    
I0902 10:03:23.842087  evented.go:295] "Received Container Started Event", container id: eacee74973c186af7e16228a0fb22bb624a25ade9c499e5355910c8659a53968, 
    and then the evented pleg got the started event of the app container eacee74973c186af7e16228a0fb22bb624a25ade9c499e5355910c8659a53968, but the second 
    container sync loop work has already started.

I0902 10:03:23.845966  kuberuntime_manager.go:968 "Container of pod is not in the desired state and shall be started" containerName="kube-apiserver" 
    pod="kube-system/kube-apiserver-kind-control-plane", so the second container sync loop work think the app container didn't started and try to create a 
    new and started failed for the first one was in running state.
```

pacoxu commented:
> the start event of the app container wasn't synced to kubelet by evented pleg timely

Does this mean that the kubelet got the started event first and then the created event? Is this expected?

wxx213 commented:
> > the start event of the app container wasn't synced to kubelet by evented pleg timely
> 
> Does this mean that the kubelet got the started event first and then the created event? Is this an expected behavior?

The created event always arrives in kubelet before started event for the same container. 

In this case, the started event arrived in kubelet after the second pod sync work started as the timeline described, so the second pod sync work could only got the created state from the pod cache when started the work(the code: [podWorkerLoop](https://github.com/kubernetes/kubernetes/blob/e5bafe2bed13fe72e88a3597930bdfbab5267e9d/pkg/kubelet/pod_workers.go#L1256)). Then the second sync work think the app container should be restarted(recreate) in [computePodActions](https://github.com/kubernetes/kubernetes/blob/e5bafe2bed13fe72e88a3597930bdfbab5267e9d/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L967)

I'm not sure why app container's started event didn't arrive in kubelet just when first pod sync worke finished where the [start](https://github.com/kubernetes/kubernetes/blob/e5bafe2bed13fe72e88a3597930bdfbab5267e9d/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1234) function already be called for the app container, maybe because the start event is send by async like [containerd](https://github.com/containerd/containerd/blob/58aa92d1c1afb86bc4012bb8f4cb49c794e3c2cc/internal/cri/server/container_start.go#L187)

Anyway, kubelet should manage this  information gap correctly with runtime.

pacoxu commented:
Returning to the current PR fix, the current solution https://github.com/kubernetes/kubernetes/pull/122778/commits/f99b04ee842d65ba979eac4f081a98cdf35e8e24 should correctly handle this scenario. Do we have some better approach? I'm afraid that this is the simplest fix for the bug.

> If a pod hangs for more than 1m, we should trigger the restart then.

IMO if the current approach is the intended fix, the key point of discussion is whether 1 minute is an appropriate time threshold to trigger the restart.

hshiina commented:
#126543 has been recently merged. A similar fix would be necessary for init containers:
https://github.com/kubernetes/kubernetes/blob/17bae91f207a5f2b101e581c90f7a72fc9adef96/pkg/kubelet/kuberuntime/kuberuntime_container.go#L1086-L1092

I reproduced the same issue with an init container on recent master:
<Details>

```
$ cat pod.yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: null
  labels:
    run: testpod
  name: testpod
spec:
  containers:
  - command:
    - sh
    - -c
    - trap exit TERM; while true; do sleep 1; done
    image: busybox
    name: testcontainer
    resources: {}
  initContainers:
  - command:
    - mkdir
    - /myvolume/test
    image: busybox
    name: testinitcontainer
    volumeMounts:
    - mountPath: /myvolume
      name: myvolume
  volumes:
  - name: myvolume
    emptyDir: {}
  dnsPolicy: ClusterFirst
  restartPolicy: Always
status: {}

$ kubectl create -f pod.yaml
pod/testpod created

$ kubectl get pod testpod -w
NAME  	READY   STATUS              	RESTARTS 	AGE
testpod   0/1 	Init:CrashLoopBackOff   2 (3s ago)   7s
testpod   0/1 	Init:0/1            	3 (14s ago)   18s
testpod   0/1 	Init:Error          	3 (14s ago)   18s
testpod   0/1 	Init:CrashLoopBackOff   3 (0s ago)	18s
testpod   0/1 	Init:0/1            	4 (27s ago)   45s
testpod   0/1 	Init:Error          	4 (27s ago)   45s
testpod   0/1 	Init:CrashLoopBackOff   4 (0s ago)	45s
```

This message implies that another init container was created:
```
$ kubectl logs testpod -c testinitcontainer
mkdir: can't create directory '/myvolume/test': File exists
```
</Details>

pacoxu commented:
/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e

pacoxu commented:
> #126543 has been recently merged. A similar fix would be necessary for init containers:
> 
> I reproduced the same issue with an init container on recent master:

@hshiina I rebased with the latest code.

Do you know if we have an e2e test to address the init container restart problem? Or probably, we should add one.

pacoxu commented:
/test pull-kubernetes-e2e-kind-beta-features
/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e
/test pull-kubernetes-e2e-kind-evented-pleg
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-features
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

hshiina commented:
> > #126543 has been recently merged. A similar fix would be necessary for init containers:
> > I reproduced the same issue with an init container on recent master:
> 
> @hshiina I rebased with the latest code.
> 
> Do you know if we have an e2e test to address the init container restart problem? Or probably, we should add one.

There doesn't look to be a test that address this problem explicitly. However, I guess a test that runs a web server on a init container like the following one will fail if we hit this issue. Init containers created later will fail to start with an error like "address already in use" and its restart count will be increased:
https://github.com/kubernetes/kubernetes/blob/8e3adc4df64d5b382c8916610313ce25e0df8e28/test/e2e/common/node/container_probe.go#L787-L788

Please note that #124297 made this issue less likely to occur.

pacoxu commented:
/test pull-kubernetes-e2e-kind-beta-features
/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e
/test pull-kubernetes-e2e-kind-evented-pleg
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-features
/test pull-kubernetes-cos-cgroupv1-containerd-node-e2e-features

hshiina committed:
EventedPLEG: Update global cache timestamp more frequently

This fix makes the period to update the global cache timestamp for
Evented PLEG shorter in order to unblock pod workers that get stuck
in `cache.GetNewerThan()` as fast as Generic PLEG. Before updating the
global timestamp, this fix calls Version() CRI API in order to confirm
there is no problem with communicating with the runtime. By this, we
can assume pods that have not been receiving events are in the latest
status.

pacoxu commented:
/test pull-kubernetes-node-e2e-containerd-alpha-features

SergeyKanzhelev commented:
should it also be handled here: https://github.com/kubernetes/kubernetes/pull/126543/? I remember that PR was introducing something similar for sidecars

This is one of the PRs that are hard to reason about - it fixes EventedPLEG feature's bug and introduces changes all over, potentially affecting generic PLEG as well. Please make sure all related places are well-documented that those are only for EventedPLEG.

Finally, is the test change covers this fix so we will catch regression in the future? Or we may need to do something more than we already have?

pacoxu commented:
> should it also be handled here: #126543? I remember that PR was introducing something similar for sidecars
> 
> This is one of the PRs that are hard to reason about - it fixes EventedPLEG feature's bug and introduces changes all over, potentially affecting generic PLEG as well. Please make sure all related places are well-documented that those are only for EventedPLEG.


Yes. This PR originally only fixed the problem of static pod when EventedPLEG is enabled. 

#126543 mentions that there is a similar issue for sidecar/init container.

> Finally, is the test change covers this fix so we will catch regression in the future? Or we may need to do something more than we already have?

This can be mocked after https://github.com/kubernetes/kubernetes/pull/127495 was merged. I can write a new Test to cover those cases.

pacoxu commented:
/test pull-kubernetes-node-e2e-containerd-standalone-mode-all-alpha

pacoxu commented:
/test pull-kubernetes-node-e2e-cri-proxy-serial

pacoxu commented:
/test pull-kubernetes-node-e2e-cri-proxy-serial

pacoxu committed:
restart sidecar container like static pod created state container

pacoxu commented:
/test pull-kubernetes-node-e2e-cri-proxy-serial

pacoxu commented:
/test pull-kubernetes-node-e2e-cri-proxy-serial

pacoxu commented:
/test pull-kubernetes-node-e2e-cri-proxy-serial

pacoxu commented:
/test pull-kubernetes-node-e2e-cri-proxy-serial

pacoxu commented:
/test pull-kubernetes-node-e2e-cri-proxy-serial

pacoxu commented:
/test pull-kubernetes-node-e2e-containerd-alpha-features
/test pull-kubernetes-node-e2e-containerd-standalone-mode-all-alpha
/test pull-crio-cgroupv1-node-e2e-eviction
/test pull-crio-cgroupv2-node-e2e-eviction
/test pull-kubernetes-node-e2e-cri-proxy-serial

/test pull-kubernetes-e2e-kind-beta-features
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e

pacoxu commented:
/test pull-kubernetes-node-e2e-containerd-alpha-features
/test pull-kubernetes-node-e2e-containerd-standalone-mode-all-alpha
/test pull-crio-cgroupv1-node-e2e-eviction
/test pull-crio-cgroupv2-node-e2e-eviction
/test pull-kubernetes-node-e2e-cri-proxy-serial

/test pull-kubernetes-e2e-kind-beta-features
/test pull-kubernetes-e2e-kind-alpha-beta-features
/test pull-kubernetes-e2e-kind-alpha-features
/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e

/test pull-kubernetes-node-crio-cgrpv1-evented-pleg-e2e-kubetest2

HirazawaUi commented:
/test pull-kubernetes-e2e-kind-evented-pleg
